pkg/kvstore/allocator/allocator.go

```diff
@@ -702,30 +704,39 @@ func (a *Allocator) RunGC() error {
 
 		// fetch list of all /value/<key> keys
 		valueKeyPrefix := path.Join(a.valuePrefix, string(v.Data))
-		k, v, err := kvstore.GetPrefix(context.Background(), valueKeyPrefix)
+		k, kvstoreValue, err := kvstore.GetPrefix(context.Background(), valueKeyPrefix)
 		if err != nil {
 			log.WithError(err).WithField(fieldPrefix, valueKeyPrefix).Warning("allocator garbage collector was unable to list keys")
 			lock.Unlock()
 			continue
 		}
 
-		// if ID has no user, delete it
-		if k == "" && v == nil {
+		// if ID has no user, mark it to be delete it in the next RunGC
+		if k == "" && kvstoreValue == nil {
 			scopedLog := log.WithFields(logrus.Fields{
 				fieldKey: key,
 				fieldID:  path.Base(key),
 			})
-			if err := kvstore.Delete(key); err != nil {
-				scopedLog.WithError(err).Warning("Unable to delete unused allocator master key")
+			// Only delete if this key was previously marked as to be deleted
+			if modRev, ok := keysToDelete[key]; ok && modRev == v.ModRevision {
+				if err := kvstore.Delete(key); err != nil {
+					scopedLog.WithError(err).Warning("Unable to delete unused allocator master key")
+				} else {
+					delete(keysToDelete, key)
+					scopedLog.Info("Deleted unused allocator master key")
+				}
 			} else {
-				scopedLog.Info("Deleted unused allocator master key")
+				// If the key was not found mark it to be delete in the next RunGC
+				keysToDeleteNew[key] = v.ModRevision
 			}
+		} else {
+			delete(keysToDelete, key)
```

Comment:
What is the purpose of this? You are returning a new map anyway.

Reply:
in a draft I was not returning a new map.